From: Jason Rhinelander <jason@imaginary.ca>
Date: Mon, 19 Oct 2026 13:20:37 -0300
Subject: Fix zmq API leakage into public API

`zmq::error_t` is essentially an internal type that we don't export,
thus causing an undefined symbol error at load time in packages that try
to use this (currently only pyoxenmq as far as I can see).  Since
`zmq::error_t` is derived from std::exception (and `what()` is virtual),
we can simply provide that rather than the zmq wrapper to avoid the
issue.
---
 oxenmq/oxenmq.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/oxenmq/oxenmq.h b/oxenmq/oxenmq.h
index a850412..48d1865 100644
--- a/oxenmq/oxenmq.h
+++ b/oxenmq/oxenmq.h
@@ -45,6 +45,7 @@
 #include <cstdint>
 #include <future>
 #include <variant>
+#include <exception>
 #include "zmq.hpp"
 #include "address.h"
 #include <oxenc/bt_serialize.h>
@@ -1476,7 +1477,7 @@ struct request_timeout {
 /// cases (such as when a message is queued but the connection fails before delivery) that do not
 /// trigger this failure at all; rather this callback only signals an immediate queuing failure.
 struct queue_failure {
-    using callback_t = std::function<void(const zmq::error_t* exc)>;
+    using callback_t = std::function<void(const std::exception* exc)>;
     /// Callback; invoked with nullptr for a queue full failure, otherwise will be set to a copy of
     /// the raised exception.
     callback_t callback;
